hilti/toolchain/src/base/preprocessor.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020-2021 by the Zeek Project. See LICENSE for details.
+
+#include <hilti/rt/util.h>
+
+#include <hilti/base/preprocessor.h>
+
+#include "base/util.h"
+
+using namespace hilti::util;
+
+hilti::Result<hilti::util::SourceCodePreprocessor::State> hilti::util::SourceCodePreprocessor::processLine(
+    const std::string_view& directive, const std::string_view& expression) {
+    if ( directive == "@if" ) {
+        if ( state() == State::Include ) {
+            auto result = _parseIf(expression);
+            if ( ! result )
+                return result.error();
+
+            _stack.push_back(*result ? 1 : 0);
+        }
+        else
+            _stack.push_back(-1);
+    }
+
+    else if ( directive == "@else" ) {
+        if ( expression.size() )
+            return result::Error("syntax error in @else directive");
+
+        if ( _stack.size() == 1 )
+            return result::Error("@else without @if");
+
+        if ( auto x = _stack.back(); x >= 0 ) {
+            _stack.pop_back();
+            _stack.push_back(1 - x);
+        }
+    }
+
+    else if ( directive == "@endif" ) {
+        if ( expression.size() )
+            return result::Error("syntax error in @else directive");
+
+        if ( _stack.size() == 1 )
+            return result::Error("@endif without @if");
+
+        _stack.pop_back();
+    }
+
+    else
+        return result::Error("unknown preprocessor directive");
+
+    return state();
+}
+
+
+hilti::Result<bool> hilti::util::SourceCodePreprocessor::_parseIf(const std::string_view& expression) {
+    bool negate = false;
+
+    auto m = hilti::rt::split(expression);
+
+    if ( m.size() >= 1 && m[0] == "!" ) {
+        negate = true;
+        m = util::slice(m, 1); // "shift m"
+    }
+
+    if ( m.size() != 1 && m.size() != 3 )
+        return result::Error("syntax error in @if directive");
+
+    auto id = std::string(m[0]);
+
+    std::string op;
+    int want = 0;
+
+    if ( m.size() == 3 ) {
+        // "<id> <operator> <expr>"
+        op = std::string(m[1]);
+        want = 0;
+        hilti::rt::atoi_n(m[2].begin(), m[2].end(), 10, &want);
```

Comment:
Not a new issue, but we probably should check the return value here to catch parse errors.